### PR TITLE
Add `Trainer(gradient_clip_algorithm='value'|'norm')`

### LINKED ISSUE #4927: clip_gradient with clip_grad_value
## 🚀 Feature
The current clip_gradient uses clip_grad_norm; can we add clip_grad_value?

### LINKED ISSUE #5456: clip_gradient with clip_grad_value
## 🚀 Feature

Same issue with #4927 
The current clip_gradient uses clip_grad_norm; can we add clip_grad_value?

https://github.com/PyTorchLightning/pytorch-lightning/blob/f2e99d617f05ec65fded81ccc6d0d59807c47573/pytorch_lightning/plugins/native_amp.py#L63-L65

### LINKED ISSUE #5460: clip_gradient with clip_grad_value
## 🚀 Feature

Same issue with #4927 #5456 

The current clip_gradient uses clip_grad_norm; can we add clip_grad_value?

https://github.com/PyTorchLightning/pytorch-lightning/blob/f2e99d617f05ec65fded81ccc6d0d59807c47573/pytorch_lightning/plugins/native_amp.py#L63-L65

============================================================
@tchaton 

As far as I know, there is a difference between clip_grad_by_value and clip_grad_by_norm.
All of the implementations in PL only use `clip_grad_by_norm`.
`clip_grad_by_value` does not perform clipping with `norm value` but just performs clipping by value, so it is useful when learning model with noisy data.
Please let me know if you think I'm wrong.

[pytorch clip by norm link](https://pytorch.org/docs/stable/generated/torch.nn.utils.clip_grad_norm_.html)
[pytorch clip by value link](https://pytorch.org/docs/stable/generated/torch.nn.utils.clip_grad_value_.html#torch.nn.utils.clip_grad_value_)

Sincerely,
Anthony Kim.

### PR DESCRIPTION
**Pull request by dhkim0225**
## What does this PR do?

Add "clip_gradients_by_value"
Fixes #5460
fixes #5456
fixes #4927
Previous PR: https://github.com/PyTorchLightning/pytorch-lightning/pull/5477
@tchaton Thanks for giving me a lot of advice.

## Before submitting
- [x] Was this discussed/approved via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [x] Did you make sure to update the documentation with your changes? (if necessary)
- [x] Did you write any new necessary tests? (not for typos and docs)
- [x] Did you verify new and existing tests pass locally with your changes?
- [x] Did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review
Anyone in the community is free to review the PR once the tests have passed.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

 - [x] Is this pull request ready for review? (if not, please submit in draft mode)
 - [x] Check that all items from **Before submitting** are resolved
 - [x] Make sure the title is self-explanatory and the description concisely explains the PR
 - [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

**Comment by Borda** [comment]
> Related Issues: #5460, #5456, #4927

what does it mean that it solves these issues?

File: CHANGELOG.md
Comment (Borda) [inline CHANGELOG.md:11]:
```suggestion

Added `gradient_clip_algorithm` argument to Trainer for gradient clipping by value ([#6123](https://github.com/PyTorchLightning/pytorch-lightning/pull/6123)).

```

File: pytorch_lightning/accelerators/accelerator.py
Comment (Borda) [inline pytorch_lightning/accelerators/accelerator.py:291]: mind format with yapf, run in terminal `yapf . -rip`

File: pytorch_lightning/plugins/precision/deepspeed_precision.py
Comment (Borda) [inline pytorch_lightning/plugins/precision/deepspeed_precision.py:63]: is the float needed?
```suggestion
        norm_type: float = 2.0,
```

File: docs/source/advanced/training_tricks.rst
Comment (Borda) [inline docs/source/advanced/training_tricks.rst:45]: can we list or refer to the code what are the types, maybe just link to `GradClipAlgorithmType`

File: pytorch_lightning/utilities/enums.py
Comment (Borda) [inline pytorch_lightning/utilities/enums.py:92]:
```suggestion
```
rather link what it is used for

File: docs/source/advanced/training_tricks.rst
Comment (dhkim0225) [reply docs/source/advanced/training_tricks.rst:46]: Good idea. I changed the documentation, but don't know the way I did is correct.

File: pytorch_lightning/accelerators/accelerator.py
Comment (dhkim0225) [reply pytorch_lightning/accelerators/accelerator.py]: updated.

**Comment by dhkim0225** [comment]
@Borda Thank you for your comments. 
I changed codes based on comments.

> > Related Issues: #5460, #5456, #4927
> 
> what does it mean that it solves these issues?

It means, this PR adds a `clip_gradient by value` option.

File: pytorch_lightning/plugins/precision/precision_plugin.py
Comment (ananthsub) [inline pytorch_lightning/plugins/precision/precision_plugin.py:111]: with the `if/else`, it's very easy to miss support in a plugin should a new `gradient_clip_algorithm` value be added.

this might be overkill, but should each precision plugin maintain a `Dict[GradClipAlgorithmType, Callable]` ? at least that will fail loudly should we try looking up that `gradient_clip_algorithm` in the dict

Comment (dhkim0225) [reply pytorch_lightning/plugins/precision/precision_plugin.py:111]: So, you mean, managing `Dict[GradClipAlgorithmType, Callable]` as an attribute of every `PrecisionPlugin`, and making two callable function, `clip_grad_by_value`  and `clip_grad_by_norm` would be better.

I agree with you. Please check what I told you is the same as your comment. I'll be work in after your confirmation.

Comment (dhkim0225) [reply pytorch_lightning/plugins/precision/precision_plugin.py:111]: Any new opinions from other reviewers?

Comment (ananthsub) [reply pytorch_lightning/plugins/precision/precision_plugin.py]: @dhkim0225 correct that's what i meant

File: pytorch_lightning/trainer/trainer.py
Comment (ananthsub) [inline pytorch_lightning/trainer/trainer.py:200]:
```suggestion
            gradient_clip_algorithm: 'value' means clip_by_value, 'norm' means clip_by_norm. Default: 'norm'
```

**Review by ananthsub — APPROVED** [review]

File: pytorch_lightning/trainer/trainer.py
Comment (dhkim0225) [reply pytorch_lightning/trainer/trainer.py:200]: All modifications have been completed based on comments. Can you check the last changes? 
https://github.com/PyTorchLightning/pytorch-lightning/pull/6123/commits/e92ec69bc7af1d6c2694141617bed9e07d4d1a67
Thanks.

File: benchmarks/test_sharded_parity.py
Comment (tchaton) [inline benchmarks/test_sharded_parity.py:123]: Wow ! You might be the first one to add a `DDPLauncher` for a test :)

Also, update the skipif with our RunIf

@RunIf(special=True, fairscale=True, min_gpus=2)

File: docs/source/advanced/training_tricks.rst
Comment (tchaton) [inline docs/source/advanced/training_tricks.rst:32]:
```suggestion
If ``gradient_clip_algorithm`` option is set to ``value``, which is ``norm`` by default, this will
`clip the gradient value <https://pytorch.org/docs/stable/nn.html#torch.nn.utils.clip_grad_value_>`_ for each parameter instead.
```

File: pytorch_lightning/accelerators/accelerator.py
Comment (tchaton) [inline pytorch_lightning/accelerators/accelerator.py:306]: Really neat !

File: pytorch_lightning/plugins/precision/precision_plugin.py
Comment (tchaton) [inline pytorch_lightning/plugins/precision/precision_plugin.py:119]: Should we make them private ?

**Review by tchaton — APPROVED** [review]
The PR looks really good ! Awesome work. Small nits to address.

File: pytorch_lightning/plugins/precision/precision_plugin.py
Comment (ananthsub) [reply pytorch_lightning/plugins/precision/precision_plugin.py:119]: @tchaton would we expect any precision plugins to override these? if not, i agree we should make them private

File: pytorch_lightning/plugins/precision/sharded_native_amp.py
Comment (ananthsub) [inline pytorch_lightning/plugins/precision/sharded_native_amp.py:37]: - here you could update `clip_grad_funcs` with the new definition for `clip_grad_by_norm` only, since after L32, the property is set. this way in case there are new grad clip algorithm types added in the base, we'll automatically get them here
- on the other hand, if we think that precision plugins will be so specialized that we should explicitly opt-in to different grad clip algorithm types, then this set up is fine

Comment (dhkim0225) [reply pytorch_lightning/plugins/precision/sharded_native_amp.py:37]: Thanks for your comment! Let's erase it. We can use `redeclaration` after we need it later.

File: benchmarks/test_sharded_parity.py
Comment (dhkim0225) [reply benchmarks/test_sharded_parity.py:120]: Adapted

File: pytorch_lightning/utilities/enums.py
Comment (carmocca) [inline pytorch_lightning/utilities/enums.py:101]: Can you expand this docstring with the difference between types?

File: tests/trainer/test_trainer.py
Comment (carmocca) [inline tests/trainer/test_trainer.py:889]: Can you use `BoringModel` instead? We are moving away from `EvalModelTemplate` because it's slower

File: tests/trainer/test_trainer.py
Comment (carmocca) [inline tests/trainer/test_trainer.py:902]: can we run this check inside one of the `LightningModule` hooks instead? In case this function ever changes, I'd rather avoid having to update this test

Comment (dhkim0225) [reply tests/trainer/test_trainer.py:902]: https://github.com/dhkim0225/pytorch-lightning/blob/2e933d4aea428b54ed90198b795602bfb3ae7ab3/tests/trainer/test_trainer.py#L847-L880

Right above there is a gradient clip_by_norm test, isn't it weird to put only clip_by_value differently?

File: pytorch_lightning/trainer/connectors/training_trick_connector.py
Comment (dhkim0225) [reply pytorch_lightning/trainer/connectors/training_trick_connector.py]: Thank you for your suggestion

File: pytorch_lightning/utilities/enums.py
Comment (dhkim0225) [reply pytorch_lightning/utilities/enums.py:101]: Okay I will.

File: tests/trainer/test_trainer.py
Comment (dhkim0225) [reply tests/trainer/test_trainer.py:889]: WIP

File: tests/trainer/test_trainer.py
Comment (carmocca) [reply tests/trainer/test_trainer.py:902]: Didn't see that one. Feel free to update it too

I think it's cleaner to do it in a public hook instead of monkey-patching the loop. It can also serve as documentation by example :)

File: pytorch_lightning/utilities/enums.py
Comment (dhkim0225) [reply pytorch_lightning/utilities/enums.py]: Updated :)

File: tests/trainer/test_trainer.py
Comment (dhkim0225) [reply tests/trainer/test_trainer.py:902]: I have read the [hooks.py](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/pytorch_lightning/core/hooks.py), but actually, I have no idea how to test it in "a public hook".
Could you please explain a little bit more?
Thanks :]

Comment (carmocca) [reply tests/trainer/test_trainer.py:902]: nvm, we don't currently have hooks to easily do this. I'll revisit it in a separate PR.

thanks!

**Review by carmocca — APPROVED** [review]

**Comment by carmocca** [comment]
@dhkim0225 I merged master into the PR but some tests are failing. Can you have a look?

File: tests/trainer/test_trainer.py
Comment (ananthsub) [inline tests/trainer/test_trainer.py:981]: please use the `BoringModel` or one of its variants instead of the `EvalModelTemplate`

**Comment by dhkim0225** [comment]
@carmocca `clip gradients by value` seems doesn't work with `horovod-cpu`.  
I keep getting an error from `horovod-cpu tests`, but it works well after I remove the `clip_grad_by_value` test in `horovod-cpu tests`.  
I'm not sure if this is normal. If it's expected behavior, we can mention in the document that `clip_gradients` does not work with horovod cpu.

**Review by ananthsub — APPROVED** [review]

**Comment by ananthsub** [comment]
`DoublePrecisionPlugin` needs to be updated too

**Comment by dhkim0225** [comment]
@ananthsub Patched :]

**Comment by kaushikb11** [comment]
@dhkim0225 Could you please address the failing [tests](https://dev.azure.com/PytorchLightning/pytorch-lightning/_build/results?buildId=4222&view=logs&j=cb44143e-1b72-550b-97cf-26ffc4f8d032&t=4a7afa63-5eac-50b7-7a3c-6e0219707934)?

**Comment by dhkim0225** [comment]
Bugs in the test codes `clip gradients by value with horovod` are fixed.
Some tests are failed since pytorch v1.8.1 released (https://github.com/pytorch/pytorch/releases/tag/v1.8.1)

File: pytorch_lightning/plugins/precision/tpu_bfloat.py
Comment (carmocca) [inline pytorch_lightning/plugins/precision/tpu_bfloat.py:31]: Why is this necessary?

File: pytorch_lightning/plugins/precision/mixed.py
Comment (carmocca) [inline pytorch_lightning/plugins/precision/mixed.py:30]: same

File: pytorch_lightning/plugins/precision/tpu_bfloat.py
Comment (dhkim0225) [reply pytorch_lightning/plugins/precision/tpu_bfloat.py:31]:
```
class PrecisionPlugin(Plugin):
    """ Plugin handling the precision-specific parts of the training.
    The static class attributes EPSILON and precision must be overwritten in child-classes and their
    default values reflect fp32 training.
    """
    EPSILON: float = 1e-6
    precision: Union[str, int] = 32

    def __init__(self) -> None:
        super().__init__()
        self.clip_grad_funcs = {
            GradClipAlgorithmType.VALUE: self.clip_grad_by_value,
            GradClipAlgorithmType.NORM: self.clip_grad_by_norm,
        }
```

`clip_gradient` functions use `self.clip_grad_funcs`. If we do not call `super().__init__()`, `self.clip_grad_funcs`is not initialized.

File: pytorch_lightning/plugins/precision/mixed.py
Comment (dhkim0225) [reply pytorch_lightning/plugins/precision/mixed.py:30]: same :]

File: pytorch_lightning/plugins/precision/tpu_bfloat.py
Comment (carmocca) [reply pytorch_lightning/plugins/precision/tpu_bfloat.py:31]: But Python adds this empty `__init__` by default so it's redundant

Comment (dhkim0225) [reply pytorch_lightning/plugins/precision/tpu_bfloat.py:31]: You're right. WIP

Comment (dhkim0225) [reply pytorch_lightning/plugins/precision/tpu_bfloat.py]: Finished !

File: pytorch_lightning/plugins/precision/mixed.py
Comment (dhkim0225) [reply pytorch_lightning/plugins/precision/mixed.py]: Changed !

File: tests/models/test_horovod.py
Comment (carmocca) [inline tests/models/test_horovod.py:53]:
```suggestion
    # append = '-a' if '.coverage' in os.listdir(_PROJECT_ROOT) else ''
```

**Comment by dhkim0225** [comment]
@carmocca Honestly, I'm tired of the slow progress.  
The first time this PR was written was on January 12th. (https://github.com/PyTorchLightning/pytorch-lightning/pull/5477)  
Merge is continuously postponed and new conflicts continue to occur, although It is a simple feature.  
Please clearly guide me if there is anything I can do to make this feature merge quickly.

File: pytorch_lightning/plugins/precision/precision_plugin.py
Comment (Borda) [reply pytorch_lightning/plugins/precision/precision_plugin.py:127]: or protected :]

**Review by Borda — APPROVED** [review]

**Comment by carmocca** [comment]
Hey @dhkim0225! Thanks a lot for your patience. This PR is just waiting for the final approval so I will just push it internally 👍

**Comment by williamFalcon** [comment]
merged!
sorry about that! Looks like this one fell through the cracks.

Thanks for the contribution!

**Pull request merged**
